### PR TITLE
Doublets on GPU

### PR DESCRIPTION
**Pull request by VinInn**
With this PR doublets (actually CACells) are created on GPU and fed to CA.
the whole workflow up to quadruplets candidates is now fully on GPU.

besides various goodies it sports a rechit map based on "first digi" that can be useful in many other cases...

some re-engineering will be welcome.
some printouts need to be removed.

cut tuning can be beneficial (although not essential)

results with ttbar PU50 realistic

image

does not Merge ok...
will be redone....

### CODE REVIEW

File: RecoPixelVertexing/PixelTriplets/plugins/gpuPixelDoublets.h [hunk 1]
```diff
@@ -11,151 +11,162 @@
...
+  void doubletsFromHisto(uint8_t const * layerPairs, uint32_t nPairs, GPUCACell * cells, uint32_t * nCells,
+                         int16_t const * iphi, Hist const * hist, uint32_t const * offsets,
+                         siPixelRecHitsHeterogeneousProduct::HitsOnGPU const & hh,
+                         GPU::VecArray< unsigned int, 2048>  * isOuterHitOfCell,
+                         int16_t const * phicuts, float const * minz, float const * maxz, float const * maxr) {
+
+    auto layerSize = [=](uint8_t li) { return offsets[li+1]-offsets[li]; };
+
+    // to be optimized later
+    uint32_t innerLayerCumulaliveSize[64];
```
Comment (makortel) [inline RecoPixelVertexing/PixelTriplets/plugins/gpuPixelDoublets.h]: Maybe fix the typo -> `innerLayerCumulativeSize`?

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletHeterogeneousEDProducer.cc [hunk 2]
```diff
@@ -98,45 +98,42 @@ void CAHitNtupletHeterogeneousEDProducer::beginStreamGPUCuda(
 void CAHitNtupletHeterogeneousEDProducer::acquireGPUCuda(
     const edm::HeterogeneousEvent &iEvent, const edm::EventSetup &iSetup,
     cuda::stream_t<> &cudaStream) {
+
+  seedingHitSets_ = std::make_unique<RegionsSeedingHitSets>();
+
+
+  // FIXME: move directly to region or similar...
   edm::Handle<IntermediateHitDoublets> hdoublets;
```
Comment (makortel) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletHeterogeneousEDProducer.cc]: I think I don't have anything else to comment (in addition to the list in the description), except "Yes!".

As discussed with @felicepantaleo and @fwyzard earlier, for the GPU workflow we definitively need to drop the "CPU doublets" (and so this `get`), and also move the "seeding layer" configuration as part of the GPU CA producer to drop the `SeedingLayersEDProducer` from the sequence.

**Comment by VinInn** [comment]
ok, after the merge nothing works anymore...

**Comment by fwyzard** [comment]
Replaced by #118.

**Pull request closed without merging**